### PR TITLE
Adding Deployment API Objects to /expapi

### PR DESCRIPTION
**Pull request by nikhiljindal**
Required for implementing: https://github.com/kubernetes/kubernetes/issues/1743
API Objects are as per https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deployment.md

cc @bgrant0607 @smarterclayton @ironcladlou

### CODE REVIEW

File: pkg/expapi/v1/types.go [hunk 1]
```diff
@@ -120,3 +120,101 @@ type HorizontalPodAutoscalerList struct {
...
+	// Users can set this to an empty string to indicate that the system should
+	// not add any selector and label. If unspecified, system uses
+	// "deployment.kubernetes.io/podTemplateHash".
+	// Value of this key is hash of DeploymentSpec.PodTemplateSpec.
+	UniqueLabelKey *string `json:"uniqueLabel,omitempty" description:"key of the label that is added to existing pods to distinguish them from new ones; deployment.kubernetes.io/podTemplateHash is used by default; value of this label is hash of pod template spec; no label is added if this is set to empty string"`
+}
+
+type DeploymentStrategy struct {
+	// Type of deployment. Can be "Recreate" or "RollingUpdate".
+	Type DeploymentType `json:"type,omitempty" description:"type of deployment; can be Recreate or RollingUpdate"`
```
Comment (bgrant0607) [inline pkg/expapi/v1/types.go]: defaults to RollingUpdate

File: pkg/expapi/v1/types.go [hunk 2]
```diff
@@ -120,3 +120,101 @@ type HorizontalPodAutoscalerList struct {
...
+
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// Rolling update config params. Present only if DeploymentType =
+	// RollingUpdate.
+	RollingUpdateSpec *RollingUpdateDeploymentSpec `json:"rollingUpdateSpec,omitempty" description:"rolling update config params; present only if deploymentType=RollingUpdate"`
```
Comment (bgrant0607) [inline pkg/expapi/v1/types.go]: json field should be called rollingUpdate

File: pkg/expapi/v1/types.go [hunk 3]
```diff
@@ -120,3 +120,101 @@ type HorizontalPodAutoscalerList struct {
...
+const (
+	// Kill all existing pods before creating new ones.
+	DeploymentRecreate DeploymentType = "Recreate"
+
+	// Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+	DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
```
Comment (bgrant0607) [inline pkg/expapi/v1/types.go]: Drop "Spec" here, also

File: pkg/expapi/v1/types.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/expapi/v1/types.go]: Added that to comment

File: pkg/expapi/v1/types.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/expapi/v1/types.go]: Updated

File: pkg/expapi/v1/types.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/expapi/v1/types.go]: Dropped

**Comment by bgrant0607** [comment]
LGTM

**Comment by zmerlynn** [comment]
Reopening to get Shippable output.

**Comment by zmerlynn** [comment]
Needs rebase

**Pull request merged**